src/pretix/base/services/quotas.py

```diff
@@ -1,15 +1,318 @@
+import sys
+from collections import Counter, defaultdict
 from datetime import timedelta
 
 from django.conf import settings
-from django.db.models import Max, Q
+from django.db.models import Count, F, Func, Max, Q, Sum
 from django.dispatch import receiver
 from django.utils.timezone import now
 from django_scopes import scopes_disabled
 
-from pretix.base.models import Event, LogEntry
+from pretix.base.models import (
+    CartPosition, Event, LogEntry, Order, OrderPosition, Quota, Voucher,
+    WaitingListEntry,
+)
 from pretix.celery_app import app
 
-from ..signals import periodic_task
+from ..signals import periodic_task, quota_availability
+
+
+class QuotaAvailability:
+    """
+    This special object allows so tompute the availability of quotas and inspect their results.
+    Initialize, add quotas with .queue(quota), compute with .compute() and access results.
+    """
+
+    def __init__(self, count_waitinglist=True, ignore_closed=False, full_results=False, early_out=True):
```

Comment:
A little more documentation would be appreciated.

`count_waitinglist` and `ignore_closed` are self-explanatory to me, `full_results` probably too. `early_out` however didn't ring a bell right away...